### DISCUSSION TITLE
Reduce bundle size during build process?

### DISCUSSION THREAD

**Discussion by whytspace** [dcastil/tailwind-merge#185] [Ideas]
First I would like to say a **huge thanks to the maintainers of this package**. I stumbled upon the _class merging problem_ the very first time I used tailwind in a React project and was about to write my own solution when I found this one 👍 

So far I have used `clsx` (and `className` before) and always liked their low impact on performacne and bundle size. Whenever I start using a new package I do a quick test on how much it adds to the production bundle and I must say I was a unpleasantly surprised when I saw the results for this package.

Here a quick comparison:

(I am using a newly created Solidjs App for size comparison)

Using string concatination: `"p-2 " + props.className`
<img width="400" alt="test-app" src="https://user-images.githubusercontent.com/4341219/212958622-5ab511ca-391a-44cc-83d0-98c62118b29a.png">

Using clsx: `clsx("p-2", props.className)`
<img width="400" alt="test-app-clsx" src="https://user-images.githubusercontent.com/4341219/212958721-43d11f94-4d0e-46ee-996c-a7b213afe98d.png">

Using tailwind-merge: `twMerge("p-2", props.className)`
<img width="400" alt="test-app-tailwind-merge" src="https://user-images.githubusercontent.com/4341219/212958924-adcf3d89-bdf1-4994-8867-0ca58a618829.png">

---

Taking a closer look at the bundle, **most of the file size comes from the default configuration object**. It includes a lot of classes for maximum compatibility, but most of them are not used in my application.

I was wondering if there was a way to **only bundle the necessary parts of the configuration**. Similar to what the just-in-time compiler does for tailwind itself.

What do you think about that? Is that a good idea?
Is it even worth the effort? (As the bundle size of the app grows `tailwind-merge`'s configuration gets smaller compared to the total size)

**Comment by dcastil** [comment]
Hey @whytspace! 👋

I like the idea of leveraging the Tailwind CSS compiler! I think I'll ask the Tailwind maintainers whether they could provide an API for it. Thanks for the suggestion!

So far tailwind-merge is just a simple static library, so I need to include support for all possible Tailwind classes in the default config. It is also not possible for Tailwind plugins to see which Tailwind classes got built, so there currently isn't an obvious way for me to dynamically reduce the bundle size unfortunately.

If bundle size is really important to you, there are two possible solutions in the short term:

### 1. Don't use tailwind-merge

If bundle size is everything and you don't have crazy merging needs, you might get away with using the Tailwind important modifier for merging. E.g. you can override `p-2` with `!p-3`. As long as you only ever need one level of overrides, this might be just enough.

### 2. Don't import the default config

When you use `twMerge` or `extendTailwindMerge`, the default config is used and therefore lands in the bundle. But you can tree-shake the config out of the bundle by using `createTailwindMerge` instead and create your own config. `createTailwindMerge` doesn't rely on the default config at all, so its bundle size is quite small ([Bundlephobia says 1.6 kB out of the 6.3 kB](https://bundlephobia.com/package/tailwind-merge@1.8.1#:~:text=of%20individual%20exports-,createTailwindMerge,-1.6kB)).

```js
const twMerge = createTailwindMerge(() => {
    return {
        …
        classGroups: {
            // You need to define all the class groups you want to use yourself now
        }
    }
})
```

**Reply by dcastil** [reply]
Linking to discussion I opened in Tailwind repo: https://github.com/tailwindlabs/tailwindcss/discussions/10348

**Reply by whytspace** [reply]
thank @dcastil! I already read about the custom config in the docs and will give it a try for now.

I am curious what the folks at tailwind have to say about the suggestion 😀

**Reply by Rinaldo** [reply]
Hi @dcastil, another approach that wouldn't require help from the tailwind team would be to break up the default config into core plugins. Right now the default config is all or nothing, you either get all the classGroups from the default with `extendTailwindMerge` or start from scratch with `createTailwindMerge`. Since tailwind-merge already supports merging configs, would it be feasible to export configs representing all the tailwind core plugins separately? Then users could specify a subset of commonly overridden plugins to reduce their bundle size.

It's common to override things like height, width, color, and some of the flex and grid properties. I would think it's less common to want to override properties like backdrop. The only issue would be if consumers get confused on which classes twMerge supports, though if they're using any tailwind plugins that's potentially already an issue.

**Reply by dcastil** [reply]
@Rinaldo that's an excellent idea! Thanks for suggesting! 🙏 I opened a new issue for this where I'll comment further on it. https://github.com/dcastil/tailwind-merge/issues/228

**Comment by ivan-kleshnin** [comment]
Bundlephobia resource seems to be dead, since the end of 2023 😮‍💨 
We should replace a link, for example with https://bundlejs.com/?q=tailwind-merge

**Reply by dcastil** [reply]
Agree, I'm also getting 502s for some while already. 😔 Created an issue in https://github.com/dcastil/tailwind-merge/issues/397.

**Comment by TasseDeCafe** [comment]
I'm going to piggy-back on this conversation to avoid creating another thread.

I'm build a React app with Vite and Tailwind, with Shadcn as UI library. Shadcn imports tailwind-merge.

I used vite-bundle-visualizer to check the size of tailwind-merge in my bundle, and it comes out at 69kb. BundleJS says it should be around 21kb.

@dcastil, do you know where this discrepancy could be coming from? Thank you!

I haven't modified the code provided by shadcn, and this is the only use of tailwind-merge in my project:

```import { clsx, type ClassValue } from "clsx"
import { twMerge } from 'tailwind-merge'

export function cn(...inputs: ClassValue[]) {
    return twMerge(clsx(inputs))
}

**Reply by whytspace** [reply]
@TasseDeCafe Thanks for joining the discussion, but I don't think that your problem is necessarily related.

>  do you know where this discrepancy could be coming from? Thank you!

`vite-bundle-visualizer` might be showing you _wrong_ data: https://github.com/KusStar/vite-bundle-visualizer/issues/8 and https://github.com/btd/rollup-plugin-visualizer/issues/96
I just mention this, because I myself have had a similar issue with the plugin recently.

**Reply by TasseDeCafe** [reply]
Huh, that's interesting! Thank you. Is there any way to cross-check this information?

**Reply by whytspace** [reply]
Other than the lib **authors** themselves, I guess not:
- https://github.com/KusStar/vite-bundle-visualizer/issues/8#issuecomment-1702314994
- https://github.com/btd/rollup-plugin-visualizer/issues/96#issuecomment-817093773

my own learning from this was that you really can't rely on the actual numbers. the only thing I use these numbers now is to compare whether a file got smaller or bigger between two builds. The actual numbers don't really matter to me in that case.

If i really need the actual file sizes, I look at the files on disk after build

**Reply by dcastil** [reply]
@TasseDeCafe I didn't check @whytspace's links but I can imagine that the discrepancy comes from the library not being minified and therefore having a bigger size. I rely on people's bundlers taking over the job of minification since that yields better results in general.

Maybe the most accurate way to check the bundle size difference is to produce a bundle with tailwind-merge included and then another one where you replace the library with no-ops. Then you can directly compare the size differences.

I also have a CI job enabled on every PR which checks the minified and Brotli-compressed bundle size of tailwind-merge (latest one [here](https://github.com/dcastil/tailwind-merge/pull/398#issuecomment-2028943982) which is around 6 kB).

**Comment by dcastil** [comment]
Note: Mentioned in https://github.com/dcastil/tailwind-merge/discussions/413#discussioncomment-9303629

**Comment by mustafa0x** [comment]
I also created an empty vite app with just the two lines below.
- [bundle.es.js](https://github.com/user-attachments/files/24268131/bundle.es.js) was a whopping 90kb
- Minified with esbuild: 74.49kb
- Minified with terser: 26.98 kB (gzipped 8.36kb)

27kb of minified JS is a fair bit of JS, especially on low end devices.

[Rolld(own)phobia](https://rolldphobia.ssssota.dev/?i=tailwind-merge%7C%7B+twMerge+%7D) says 12.13 KiB (minified). I'm not sure why it's so much less. Edit: I copied the generated code. It's ~24kb.

**Reply by mustafa0x** [reply]
The first 20 lines are a polyfill for `<link rel="modulepreload">`. 
```javascript
function() {
    const e = document.createElement("link").relList;
    if (!(e && e.supports && e.supports("modulepreload"))) { ... }
}()
```
Modern browsers (Chrome 66+, Edge 79+, Safari 11.1+) support modules natively. Deleting this block saves ~500 bytes.

**Comment by gianpaj** [comment]
I just noticed as well, trying to optimise my website's bundle.
<img width="1400" height="667" alt="image" src="https://github.com/user-attachments/assets/66956dac-0150-42c1-b617-06edc1d54ebd" />
```
tailwind-merge/
23.56 KB compressed (estimated)
76.10 KB uncompressed
```
😮

**Reply by dcastil** [reply]
Hey @gianpaj! 👋

tailwind-merge relies on a [large config](https://github.com/dcastil/tailwind-merge/blob/v3.6.0/src/lib/default-config.ts) to figure out which Tailwind classes are in conflict with each other. `twMerge` and `extendTailwindMerge` both rely on this config.

If you want to reduce your tailwind-merge footprint, the easiest way is to use `createTailwindMerge` instead and create your own config. You can copy the default config and remove everything that you know you don't need since there are a lot of niche classes in Tailwind CSS. The two caveats here are:
1. You need to keep this config in sync with your usage because classes that aren't in the config will not get merged properly. If you ever use some class group that you removed from the config, tailwind-merge will not resolve its conflicts.
2. When tailwind-merge gets updated, e.g. for newer Tailwind CSS versions, you'll have more code to migrate.

There is also the possibility to eject out of tailwind-merge entirely if the bundle size is not worth it for you. I wrote about some viable alternatives here: https://github.com/dcastil/tailwind-merge/blob/v3.6.0/docs/when-and-how-to-use-it.md#alternatives

Btw. I track the minified tailwind-merge size in every PR, you can see it here https://github.com/dcastil/tailwind-merge/pull/697#issuecomment-4961305718. Just expand the "all size metrics" section.

**Comment by switz** [comment]
Would love to see a much slimmer tailwind merge

**Reply by dcastil** [reply]
Hey @switz! 👋 Just wrote about some options here that could be interesting to you: https://github.com/dcastil/tailwind-merge/discussions/185#discussioncomment-17762486

But I'm also always happy to hear suggestions how to make tailwind-merge's bundle size smaller. 🙂
